File: app/src/main/java/bb/incognito/view/adapter/CocktailAdapter.java
package bb.incognito.view.adapter;

import android.databinding.DataBindingUtil;
import android.graphics.Color;
import android.graphics.drawable.Drawable;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;
import android.widget.CompoundButton;
import android.widget.Filter;
import android.widget.Filterable;

import java.util.ArrayList;
import java.util.List;

import bb.incognito.R;
import bb.incognito.databinding.CocktailRowBinding;
import bb.incognito.model.Cocktail;
import bb.incognito.viewModel.CocktailRowVM;

public class CocktailAdapter extends RecyclerView.Adapter<CocktailAdapter.CocktailViewHolder> implements Filterable {
    private List<Cocktail> cocktails = new ArrayList<>();
    private List<Cocktail> filteredCocktails = new ArrayList<>();
    private CocktailAdapter.ItemFilter filter = new CocktailAdapter.ItemFilter();
    public boolean checkable = false;
    List<Cocktail> checkedCocktails = new ArrayList<>();

    public void setCheckable(boolean b){
        checkable = b;
    }

    public List<Cocktail> getCheckedCocktails(){
        return checkedCocktails;
    }

    public void setCocktailList(List<Cocktail> cocktailList) {
        cocktails = cocktailList;
        filteredCocktails = cocktailList;
        notifyDataSetChanged();
    }

    public List<Cocktail> getCocktailList() { return cocktails; }

    public void restoreItem(int removedPosition, Cocktail removedCocktail) {
        cocktails.add(removedPosition, removedCocktail);
        notifyItemInserted(removedPosition);
    }

    public void removeAt(int position) {
        filteredCocktails.remove(position);
        notifyItemRemoved(position);
    }

    @Override
    public CocktailAdapter.CocktailViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        CocktailRowBinding cardRowBinding =
                DataBindingUtil.inflate(LayoutInflater.from(parent.getContext()), R.layout.cocktail_row,
                        parent, false);
        if (checkable)
        {
            cardRowBinding.checkBox.setVisibility(View.VISIBLE);
            cardRowBinding.cocktailRow.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View view) {
                    boolean invertState = !cardRowBinding.checkBox.isChecked();
                    cardRowBinding.checkBox.setChecked(invertState);
                }
            });

        }
        return new CocktailAdapter.CocktailViewHolder(cardRowBinding);
    }

    @Override
    public void onBindViewHolder(CocktailAdapter.CocktailViewHolder holder, int position) {
        holder.bindCard(filteredCocktails.get(position));
        holder.cocktailRowBinding.checkBox.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {

            @Override
            public void onCheckedChanged(CompoundButton compoundButton, boolean b) {
                if (b)
                    checkedCocktails.add(cocktails.get(position));
                else
                    checkedCocktails.remove(position);
            }
        });
    }

    @Override
    public int getItemCount() {
        return filteredCocktails.size();
    }

    @Override
    public Filter getFilter() {
        return filter;
    }

    class CocktailViewHolder extends RecyclerView.ViewHolder {
        CocktailRowBinding cocktailRowBinding;
        Drawable settings;
        int position;

        public CocktailViewHolder(CocktailRowBinding cocktailRowBinding) {
            super(cocktailRowBinding.cocktailRow);
            this.cocktailRowBinding = cocktailRowBinding;
        }

        void bindCard(Cocktail cocktail) {
            if (cocktailRowBinding.getViewModel() == null) {
                cocktailRowBinding.setViewModel(
                        new CocktailRowVM(cocktail));
            } else {
                cocktailRowBinding.getViewModel().setCocktail(cocktail);
            }
        }
        public boolean isChecked()
        {
            return cocktailRowBinding.checkBox.isChecked();
        }

        public void onItemSelected() {
            settings = itemView.getBackground();
            itemView.setBackgroundColor(Color.LTGRAY);
        }

        public void onItemClear() {
            itemView.setBackground(settings);
        }
    }

    private class ItemFilter extends Filter {
        @Override
        protected FilterResults performFiltering(CharSequence constraint) {
            FilterResults results = new FilterResults();
            List<Cocktail> newList = new ArrayList<>(cocktails.size());

            for(Cocktail c : cocktails)
            {
                if(c.getName().toLowerCase().contains(constraint))
                    newList.add(c);
            }
            results.values = newList;
            results.count = newList.size();
            return results;
        }

        @SuppressWarnings("unchecked")
        @Override
        protected void publishResults(CharSequence constraint, FilterResults results) {
            filteredCocktails = (ArrayList<Cocktail>) results.values;
            notifyDataSetChanged();
        }
    }
}

File: app/src/main/java/com/koktajlbar/incognitobook/view/adapter/CocktailAdapter.kt
package com.koktajlbar.incognitobook.view.adapter

import android.graphics.drawable.Drawable
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Filter
import android.widget.Filterable
import androidx.databinding.DataBindingUtil
import androidx.fragment.app.FragmentActivity
import androidx.recyclerview.widget.RecyclerView
import com.koktajlbar.incognitobook.R
import com.koktajlbar.incognitobook.databinding.CocktailRowBinding
import com.koktajlbar.incognitobook.model.Cocktail
import com.koktajlbar.incognitobook.viewmodels.CocktailRowVM
import java.util.*

class CocktailAdapter(val activity: FragmentActivity?) : RecyclerView.Adapter<CocktailAdapter.CocktailViewHolder>(), Filterable {
    private var cocktails: MutableList<Cocktail> = ArrayList()
    private var filteredCocktails: MutableList<Cocktail> = ArrayList()
    private val filter = ItemFilter()
    private var checkable = false
    private var checkedCocktails: MutableList<Cocktail> = ArrayList()

    var cocktailList: MutableList<Cocktail>
        get() = cocktails
        set(cocktailList) {
            cocktails = cocktailList
            filteredCocktails = cocktailList
            notifyDataSetChanged()
        }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CocktailViewHolder {
        val cardRowBinding = DataBindingUtil.inflate<CocktailRowBinding>(LayoutInflater.from(parent.context), R.layout.cocktail_row,
                parent, false)
        if (checkable) {
            cardRowBinding.checkBox.visibility = View.VISIBLE
            cardRowBinding.cocktailRow.setOnClickListener {
                val invertState = !cardRowBinding.checkBox.isChecked
                cardRowBinding.checkBox.isChecked = invertState
            }

        }
        return CocktailViewHolder(cardRowBinding)
    }

    override fun onBindViewHolder(holder: CocktailViewHolder, position: Int) {
        holder.bindCard(filteredCocktails[position])
        holder.cocktailRowBinding.checkBox.setOnCheckedChangeListener { _, b ->
            if (b)
                checkedCocktails.add(cocktails[position])
            else
                checkedCocktails.remove(cocktails[position])
        }
    }

    override fun getItemCount(): Int {
        return filteredCocktails.size
    }

    override fun getFilter(): Filter {
        return filter
    }

    inner class CocktailViewHolder(var cocktailRowBinding: CocktailRowBinding) : RecyclerView.ViewHolder(cocktailRowBinding.cocktailRow) {
        var settings: Drawable? = null
        var position: Int? = 0

        fun bindCard(cocktail: Cocktail) {
            if (cocktailRowBinding.viewModel == null) {
                cocktailRowBinding.viewModel = CocktailRowVM(cocktail)
            } else {
                cocktailRowBinding.viewModel!!.setCocktail(cocktail)
            }
        }

    }

    private inner class ItemFilter : Filter() {
        override fun performFiltering(constraint: CharSequence): FilterResults {
            val results = FilterResults()
            val newList = ArrayList<Cocktail>(cocktails.size)

            for (c in cocktails) {
                if (c.name.toLowerCase(Locale.getDefault()).contains(constraint) || c.category.toLowerCase(Locale.getDefault()).contains((constraint)))
                    newList.add(c)
            }
            results.values = newList
            results.count = newList.size
            return results
        }

        override fun publishResults(constraint: CharSequence, results: FilterResults) {
            filteredCocktails = results.values as ArrayList<Cocktail>
            notifyDataSetChanged()
        }
    }
}
